packages/edit-navigation/src/components/menu-editor/use-navigation-block-editor.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useCallback } from '@wordpress/element';
+import { useEntityBlockEditor } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import useCreateMissingMenuItems from './use-create-missing-menu-items';
+import batchSave from './batch-save';
+
+export default function useNavigationBlockEditor( query, postId ) {
+	const [ createMissingMenuItems, onCreated ] = useCreateMissingMenuItems(
+		query
+	);
+	const saveMenuItems = useSaveMenuItems( query );
+	const save = () => onCreated( () => saveMenuItems( blocks ) );
+
+	const [ blocks, onInput, _onChange ] = useEntityBlockEditor(
+		'root',
+		'postType',
+		{ id: postId }
+	);
+	const onChange = useCallback(
+		( updatedBlocks ) => {
+			_onChange( updatedBlocks );
+			createMissingMenuItems( blocks, updatedBlocks );
+		},
+		[ blocks, onChange, createMissingMenuItems ]
+	);
+
+	return [ blocks, onInput, onChange, save ];
+}
+
+export function useSaveMenuItems( query ) {
+	const { createSuccessNotice, createErrorNotice } = useDispatch(
+		'core/notices'
+	);
+	const select = useSelect( ( s ) => s );
+	const saveBlocks = async ( blocks ) => {
+		const menuItemsByClientId = mapMenuItemsByClientId(
+			select( 'core' ).getMenuItems( query ),
+			select( 'core/edit-navigation' ).getMenuItemIdToClientIdMapping(
+				query
+			)
+		);
```

Comment:
`saveBlocks` is not called synchronously when the save button is pressed. Instead, we wait for all draft menu items to be created first, and only then `saveBlocks` is called. This means that we need a fresh version of `menuItemId -> clientId` mapping - if we simply selected it in line 42 above then it would be outdated by the time this function is called. Ditto for `getMenuItems`. I am not too happy about this solution, but the only alternative I can think of that would involve a "valid" use of `useSelect` calls is keeping a state variable like `const [shouldSave, setShouldSave] = useState( false );`. Then, when all the async operations are finished, we'd re-render the component using `setShouldSave( true );`, and react to this change in an effect which would then have access to the most recent version of `select()`-ed data.